src/events.rs

```diff
@@ -41,7 +41,7 @@ pub struct EventNotifier(Arc<NotifyFn>);
 impl EventNotifier {
     pub fn new<F, O>(notify0: F) -> Self
     where
-        F: Fn(Event) -> O + 'static,
+        F: (Fn(Event) -> O) + Send + Sync + 'static,
```

Comment:
I think this `Sync` bound is again not necessary.

Reply:
Removing that constraint from the `EventNotifier` boxed value causes this compilation error:

```
❯ cargo test
   Compiling capataz v0.0.0 (/home/roman/tmp/rust-capataz)
error[E0277]: `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)` cannot be shared between threads safely
   --> src/supervisor/mod.rs:480:30
    |
480 |           let start_notifier = StartNotifier::from_oneshot(started_send);
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)` cannot be shared between threads safely
    | 
   ::: src/notifier.rs:8:5
    |
8   | /     pub fn from_oneshot(sender: oneshot::Sender<Result<T, E>>) -> Self
9   | |     where
10  | |         T: Send + 'static,
11  | |         E: Send + 'static,
...   |
15  | |         }))
16  | |     }
    | |_____- required by `notifier::StartNotifier::<T, E>::from_oneshot`
    |
    = help: the trait `std::marker::Sync` is not implemented for `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>`
    = note: required because it appears within the type `std::boxed::Box<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::boxed::Box<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>>`
    = note: required because it appears within the type `events::EventNotifier`
    = note: required because it appears within the type `supervisor::SpecMeta`
    = note: required because it appears within the type `supervisor::Spec`
    = note: required because it appears within the type `(supervisor::Spec, std::sync::Arc<supervisor::StartError>)`

error[E0277]: `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)` cannot be shared between threads safely
   --> src/supervisor/mod.rs:488:27
    |
488 |         let join_handle = task::spawn(run_supervisor_monitor(
    |                           ^^^^^^^^^^^ `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)` cannot be shared between threads safely
    | 
   ::: /home/roman/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.22/src/task/spawn.rs:128:20
    |
128 |         T::Output: Send + 'static,
    |                    ---- required by this bound in `tokio::spawn`
    |
    = help: the trait `std::marker::Sync` is not implemented for `(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>`
    = note: required because it appears within the type `std::boxed::Box<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::boxed::Box<(dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'static)>> + std::marker::Send + 'static)>>`
    = note: required because it appears within the type `events::EventNotifier`
    = note: required because it appears within the type `supervisor::SpecMeta`
    = note: required because it appears within the type `supervisor::Spec`
    = note: required because it appears within the type `supervisor::SupervisorResult`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `capataz`.
```